app/vlinsert/datadog/datadog.go

```diff
@@ -129,19 +142,43 @@ func readLogsRequest(ts int64, data []byte, lmp insertutils.LogMessageProcessor)
 			if err != nil {
 				return
 			}
-			val, e := v.StringBytes()
-			if e != nil {
-				err = fmt.Errorf("unexpected label value type for %q:%q; want string", k, v)
-				return
-			}
-			switch string(k) {
+			switch bytesutil.ToUnsafeString(k) {
 			case "message":
-				fields = append(fields, logstorage.Field{
-					Name:  "_msg",
-					Value: bytesutil.ToUnsafeString(val),
-				})
+				switch v.Type() {
+				case fastjson.TypeString:
+					val := v.GetStringBytes()
+					fields = append(fields, logstorage.Field{
+						Name:  "_msg",
+						Value: bytesutil.ToUnsafeString(val),
+					})
+				case fastjson.TypeObject:
+					p := logstorage.GetJSONParser()
+					if e := p.ParseLogMessage(v.MarshalTo(nil)); e != nil {
+						err = fmt.Errorf("cannot parse json-encoded log entry: %w", e)
+						return
+					}
+					logstorage.RenameField(p.Fields, []string{"message"}, "_msg")
+					fields = append(fields, p.Fields...)
+					logstorage.PutJSONParser(p)
```

Comment:
given code violates `ParseLogMessage` rule. It holds reference for p.Fields after `PutJSONParser` call.